Migrate to tomland v0.5.0

Hello. I decided to tackle a couple Hacktoberfest issues here, and I encounter this error during dependencies installation.

`cabal install` produces the following output:

```
src/Toml/Parser.hs:30:33: error:
    Module ‘Text.Megaparsec’ does not export ‘parseErrorPretty'’
   |
30 | import Text.Megaparsec (Parsec, parseErrorPretty', try)
   |                                 ^^^^^^^^^^^^^^^^^

src/Toml/Parser.hs:31:44: error:
    Module ‘Text.Megaparsec.Char’ does not export ‘anyChar’
   |
31 | import Text.Megaparsec.Char (alphaNumChar, anyChar, char, digitChar, eol, hexDigitChar, oneOf,
   |                                            ^^^^^^^

src/Toml/Parser.hs:31:89: error:
    Module ‘Text.Megaparsec.Char’ does not export ‘oneOf’
   |
31 | import Text.Megaparsec.Char (alphaNumChar, anyChar, char, digitChar, eol, hexDigitChar, oneOf,
   |                                                                                         ^^^^^

src/Toml/Parser.hs:32:30: error:
    Module ‘Text.Megaparsec.Char’ does not export ‘satisfy’
   |
32 |                              satisfy, space, space1, string, tab)
   |                              ^^^^^^^
```

So I cloned the `tomland` package and installed it from source. As the version number suggests, I didn't expect backward compability and that expectation was met:

```
src/Summoner/Source.hs:10:21: error:
    Module ‘Toml’ does not export ‘BiToml’
   |
10 | import Toml (BiMap, BiToml, Key)
   |                     ^^^^^^
```

Is it just me or the `summoner` project should be updated according to the new version of `tomland`?

P. S. Yes, I checked out the `v0.4.0` tag. I didn't manage to build this version at all, `cabal install` command results in the same error as in the first output of this message.

--- a/.travis.yml
+++ b/.travis.yml
@@ -34,15 +34,15 @@ matrix:
         - ghc-8.4.4
         - cabal-install-head
 
-  - ghc: 8.6.1
-    env: GHCVER='8.6.1' CABALVER='head'
+  - ghc: 8.6.2
+    env: GHCVER='8.6.2' CABALVER='head'
     os: linux
     addons:
       apt:
         sources:
         - hvr-ghc
         packages:
-        - ghc-8.6.1
+        - ghc-8.6.2
         - cabal-install-head
 
   - ghc: 8.2.2
@@ -61,8 +61,8 @@ matrix:
         packages:
         - libgmp-dev
 
-  - ghc: 8.6.1
-    env: GHCVER='8.6.1' STACK_YAML="$TRAVIS_BUILD_DIR/stack-$GHCVER.yaml"
+  - ghc: 8.6.2
+    env: GHCVER='8.6.2' STACK_YAML="$TRAVIS_BUILD_DIR/stack-$GHCVER.yaml"
     os: linux
     addons:
       apt:

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -21,8 +21,9 @@
   Upgrade `stack` `intall-cabal` option's version to `2.2.0.1` in
   the generated Travis file.
 * [#197](https://github.com/kowainik/summoner/issues/197)
-  Build with `stack` lts 12.17, and support `GHC` 8.4.4 in
-  project generation.
+  Build with `stack` lts-12.18, and support `GHC` 8.4.4 in
+  project generation. Make GHC-8.4.4 the default one for generated project.
+  Drop support for GHC-8.4.3 on CI. Build with GHC-8.4.4 on CI.
 * [#142](https://github.com/kowainik/summoner/issues/142):
   Add version bounds to `base` in the generated `.cabal` file.
 * [#199](https://github.com/kowainik/summoner/issues/199):
@@ -34,6 +35,7 @@
   Update prompt questions for CLI
 * [#155](https://github.com/kowainik/summoner/issues/155):
   Bump up to `tomland-0.5.0`.
+* Drop support for GHC-8.6.1 on CI. Support GHC-8.6.2 on CI.
 
 1.1.0.1
 =======

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -32,6 +32,8 @@
   each license.
 * [#195](https://github.com/kowainik/summoner/issues/195):
   Update prompt questions for CLI
+* [#155](https://github.com/kowainik/summoner/issues/155):
+  Bump up to `tomland-0.5.0`.
 
 1.1.0.1
 =======

--- a/src/Summoner/Config.hs
+++ b/src/Summoner/Config.hs
@@ -26,12 +26,12 @@ module Summoner.Config
 import Data.List (lookup)
 import Generics.Deriving.Monoid (GMonoid, gmemptydefault)
 import Generics.Deriving.Semigroup (GSemigroup, gsappenddefault)
-import Toml (AnyValue (..), BiMap (..), BiToml, Bijection (..), Key, dimap, (.=))
+import Toml (BiMap (..), Key, TomlCodec, (.=))
 
 import Summoner.Decision (Decision (..))
 import Summoner.GhcVer (GhcVer (..), parseGhcVer, showGhcVer)
 import Summoner.License (LicenseName (..), parseLicenseName)
-import Summoner.Settings (CustomPrelude (..))
+import Summoner.Settings (CustomPrelude (..), customPreludeT)
 import Summoner.Source (Source, sourceT)
 import Summoner.Validation (Validation (..))
 
@@ -128,7 +128,7 @@ defaultConfig = Config
     }
 
 -- | Identifies how to read 'Config' data from the @.toml@ file.
-configT :: BiToml PartialConfig
+configT :: TomlCodec PartialConfig
 configT = Config
     <$> lastT Toml.text "owner"       .= cOwner
     <*> lastT Toml.text "fullName"    .= cFullName
@@ -145,44 +145,29 @@ configT = Config
     <*> decision        "exe"         .= cExe
     <*> decision        "test"        .= cTest
     <*> decision        "bench"       .= cBench
-    <*> lastT (Toml.table preludeT) "prelude" .= cPrelude
-    <*> textArr         "extensions"  .= cExtensions
-    <*> textArr         "warnings"    .= cWarnings
-    <*> wrapLastT (maybeSourceT "stylish") .= cStylish
-    <*> wrapLastT (maybeSourceT "contributing") .= cContributing
+    <*> lastT preludeT "prelude"      .= cPrelude
+    <*> textArr        "extensions"   .= cExtensions
+    <*> textArr        "warnings"     .= cWarnings
+    <*> lastT sourceT  "stylish"      .= cStylish
+    <*> lastT sourceT  "contributing" .= cContributing
   where
-    wrapLastT :: BiToml (Maybe a) -> BiToml (Last a)
-    wrapLastT = Toml.dimap getLast Last
+    lastT :: (Key -> TomlCodec a) -> Key -> TomlCodec (Last a)
+    lastT codec = Toml.dimap getLast Last . Toml.dioptional . codec
 
-    maybeSourceT :: Key -> BiToml (Maybe Source)
-    maybeSourceT key = dimaybeT (sourceT key)
+    _GhcVer :: BiMap GhcVer Toml.AnyValue
+    _GhcVer = Toml._TextBy showGhcVer parseGhcVer
 
-    dimaybeT :: BiToml a -> BiToml (Maybe a)
-    dimaybeT bi = Bijection
-        { biRead  = optional (biRead bi)
-        , biWrite = traverse (biWrite bi)
-        }
-
-    lastT :: (Key -> BiToml a) -> Key -> BiToml (Last a)
-    lastT = Toml.wrapper . Toml.maybeT
-
-    _GhcVer :: BiMap AnyValue GhcVer
-    _GhcVer = BiMap
-        { forward = \(AnyValue t) -> Toml.matchText t >>= parseGhcVer
-        , backward = Just . AnyValue . Toml.Text . showGhcVer
-        }
-
-    ghcVerArr :: Key -> BiToml [GhcVer]
+    ghcVerArr :: Key -> TomlCodec [GhcVer]
     ghcVerArr = Toml.arrayOf _GhcVer
 
-    license :: Key -> BiToml LicenseName
+    license :: Key -> TomlCodec LicenseName
     license = Toml.mdimap show parseLicenseName . Toml.text
 
-    textArr :: Key -> BiToml [Text]
-    textArr = dimap Just maybeToMonoid . Toml.maybeT (Toml.arrayOf Toml._Text)
+    textArr :: Key -> TomlCodec [Text]
+    textArr = Toml.dimap Just maybeToMonoid . Toml.dioptional . Toml.arrayOf Toml._Text
 
-    decision :: Key -> BiToml Decision
-    decision = dimap fromDecision toDecision . Toml.maybeT Toml.bool
+    decision :: Key -> TomlCodec Decision
+    decision = Toml.dimap fromDecision toDecision . Toml.dioptional . Toml.bool
 
     decisionMaybe :: [(Decision, Maybe Bool)]
     decisionMaybe = [ (Idk, Nothing)
@@ -196,10 +181,9 @@ configT = Config
     toDecision :: Maybe Bool -> Decision
     toDecision m = fromMaybe (error "Impossible") $ lookup m $ map swap decisionMaybe
 
-    preludeT :: BiToml CustomPrelude
-    preludeT = CustomPrelude
-        <$> Toml.text "package" .= cpPackage
-        <*> Toml.text "module"  .= cpModule
+    preludeT :: Key -> TomlCodec CustomPrelude
+    preludeT = Toml.table customPreludeT
+
 
 -- | Make sure that all the required configurations options were specified.
 finalise :: PartialConfig -> Validation [Text] Config

--- a/src/Summoner/GhcVer.hs
+++ b/src/Summoner/GhcVer.hs
@@ -44,7 +44,7 @@ latestLts = \case
     Ghc802  -> "9.21"
     Ghc822  -> "11.22"
     Ghc843  -> "12.14"
-    Ghc844  -> "12.17"
+    Ghc844  -> "12.18"
 
 -- | Represents PVP versioning (4 numbers).
 data Pvp = Pvp

--- a/src/Summoner/Settings.hs
+++ b/src/Summoner/Settings.hs
@@ -1,16 +1,26 @@
 module Summoner.Settings
        ( Settings (..)
        , CustomPrelude (..)
+       , customPreludeT
        ) where
+
+import Toml (TomlCodec, (.=))
 
 import Summoner.GhcVer (GhcVer)
 import Summoner.License (License, LicenseName)
+
+import qualified Toml
 
 
 data CustomPrelude = CustomPrelude
     { cpPackage :: Text
     , cpModule  :: Text
     } deriving (Show, Eq)
+
+customPreludeT :: TomlCodec CustomPrelude
+customPreludeT = CustomPrelude
+    <$> Toml.text "package" .= cpPackage
+    <*> Toml.text "module"  .= cpModule
 
 -- | Data needed for project creation.
 data Settings = Settings

--- a/src/Summoner/Source.hs
+++ b/src/Summoner/Source.hs
@@ -10,7 +10,7 @@ import Control.Arrow ((>>>))
 import Control.Exception (catch)
 import NeatInterpolation (text)
 import System.Process (readProcess)
-import Toml (BiMap, BiToml, Key)
+import Toml (BiMap, Key, TomlCodec)
 
 import Summoner.Ansi (errorMessage)
 
@@ -25,29 +25,36 @@ data Source
     | Link Text
     deriving (Show, Eq)
 
-sourceT :: Key -> BiToml Source
-sourceT nm = Toml.match (Toml._Text   >>> _Url)  (nm <> "url")
-         <|> Toml.match (Toml._String >>> _File) (nm <> "file")
-         <|> Toml.match (Toml._Text >>> _Link) (nm <> "link")
+matchUrl :: Source -> Maybe Text
+matchUrl (Url url) = Just url
+matchUrl _         = Nothing
+
+matchFile :: Source -> Maybe FilePath
+matchFile (File file) = Just file
+matchFile _           = Nothing
+
+matchLink :: Source -> Maybe Text
+matchLink (Link link) = Just link
+matchLink _           = Nothing
+
+sourceT :: Key -> TomlCodec Source
+sourceT nm = Toml.match (_Url  >>> Toml._Text)   (nm <> "url")
+         <|> Toml.match (_File >>> Toml._String) (nm <> "file")
+         <|> Toml.match (_Link >>> Toml._Text)   (nm <> "link")
   where
-    _Url :: BiMap Text Source
-    _Url = Toml.invert $ Toml.prism (source Just (const Nothing) (const Nothing)) Url
+    _Url :: BiMap Source Text
+    _Url = Toml.prism Url matchUrl
 
-    _File :: BiMap FilePath Source
-    _File = Toml.invert $ Toml.prism (source (const Nothing) Just (const Nothing)) File
+    _File :: BiMap Source FilePath
+    _File = Toml.prism File matchFile
 
-    _Link :: BiMap Text Source
-    _Link = Toml.invert $ Toml.prism (source (const Nothing) (const Nothing) Just) Link
-
-    source :: (Text -> c) -> (FilePath -> c) -> (Text -> c) -> Source -> c
-    source f _ _ (Url x)  = f x
-    source _ f _ (File x) = f x
-    source _ _ f (Link x) = f x
+    _Link :: BiMap Source Text
+    _Link = Toml.prism Link matchLink
 
 fetchSource :: Source -> IO (Maybe Text)
 fetchSource = \case
     File path -> catch (Just <$> readFileText path) (fileError path)
-    Url url -> catch (fetchUrl url) (urlError url)
+    Url url   -> catch (fetchUrl url) (urlError url)
     Link link -> putLink link
   where
     fileError :: FilePath -> SomeException -> IO (Maybe Text)

--- a/stack-8.2.2.yaml
+++ b/stack-8.2.2.yaml
@@ -2,7 +2,8 @@ resolver: lts-11.22
 
 extra-deps:
   - base-noprelude-4.10.1.0
-  - megaparsec-6.5.0
-  - neat-interpolation-0.3.2.2
   - relude-0.3.0
-  - tomland-0.4.0
+  - tomland-0.5.0
+  - megaparsec-7.0.4            # needed for tomland-0.5.0
+  - parser-combinators-1.0.0    # needed for megaparsec >= 7
+  - neat-interpolation-0.3.2.4  # needed because of megaparsec >= 7

--- a/stack-8.6.1.yaml
+++ b/stack-8.6.1.yaml
@@ -1,5 +1,0 @@
-resolver: nightly-2018-10-01
-
-extra-deps:
-  - base-noprelude-4.12.0.0
-  - tomland-0.4.0

--- a/stack-8.6.2.yaml
+++ b/stack-8.6.2.yaml
@@ -1,0 +1,5 @@
+resolver:  nightly-2018-11-15
+
+extra-deps:
+  - base-noprelude-4.12.0.0
+  - relude-0.3.0

--- a/stack.yaml
+++ b/stack.yaml
@@ -1,6 +1,8 @@
-resolver: lts-12.17
+resolver: lts-12.18
 
 extra-deps:
   - base-noprelude-4.11.1.0
   - relude-0.3.0
-  - tomland-0.4.0
+  - tomland-0.5.0
+  - megaparsec-7.0.4            # needed for tomland-0.5.0
+  - neat-interpolation-0.3.2.4  # needed because of megaparsec >= 7

--- a/summoner.cabal
+++ b/summoner.cabal
@@ -18,7 +18,7 @@ extra-doc-files:     README.md
                    , CHANGELOG.md
 tested-with:         GHC == 8.2.2
                    , GHC == 8.4.4
-                   , GHC == 8.6.1
+                   , GHC == 8.6.2
 
 source-repository head
   type:     git
@@ -67,7 +67,7 @@ library
                      , relude ^>= 0.3.0
                      , text ^>= 1.2.3.0
                      , time ^>= 1.8
-                     , tomland ^>= 0.4.0
+                     , tomland ^>= 0.5.0
 
   ghc-options:         -Wall
                        -Wincomplete-uni-patterns

